[pkg/ottl] Added 'Adding new functions/converters' guidelines (#33117)
open-telemetry/opentelemetry-collector-contrib: pkg/ottl/ottlfuncs/README.md

- Functions that interact with multiple items MUST have plurality in the name. Ex: `truncate_all`, `keep_keys`, `replace_all_matches`.
- Functions that interact with a single item MUST NOT have plurality in the name. If a function would interact with multiple items due to a condition, like `where`, it is still considered singular. Ex: `set`, `delete`, `replace_match`.
- Functions that change a specific target MUST set the target as the first parameter.

> ## Adding New Functions/Converters

Reviewer: ```suggestion
## Adding New Editors/Converters
```

> ## Adding New Functions/Converters
> 
> Before raising a PR with a new function or converter, raise a PR to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.

Reviewer: ```suggestion
## Adding New Editors/Converters

Before raising a PR with a new Editor or Converter, raise an issue to describe the proposed function. While acceptance strongly depends on factors specific to each proposal, consider these guidelines for early assessment.
```

> You're more than welcome to open an issue if:

Reviewer 2: ```suggestion
Your proposal likely will be accepted if:
```

- The proposed functionality is missing,
- The proposed solution significantly improves user experience and readability for very common use cases,
- The proposed solution is more performant in cases where it is possible to achieve the same result with existing options.

It will be up for discussion if your proposal solves an issue that can be achieved in another way but does not improve user experience or performance.

Your proposal likely won't be accepted if:
- User experience is worse and assumes a highly technical user,
- The performance of your proposal very negatively affects the processing pipeline.

As with code, `ottl` should aim for readability first. This means:
- Using short, meaningful, and descriptive names,

> - Ensuring naming consistency across functions and converters,

Reviewer: ```suggestion
- Ensuring naming consistency across Editors and Converters,
```

- Avoiding deep nesting to achieve desired transformations,

> - Ensuring functions and converters have a single responsibility,

Reviewer: ```suggestion
- Ensuring Editors and Converters have a single responsibility,
```

> - Following the DRY principle, avoiding multiple functions doing the same thing.

Reviewer 2: I think we can omit this principle considering we state above how important user experience is and we are gonna encounter lots of scenarios where `replace_pattern` could be a solution, but isn't user-friendly.

- Functions that interact with multiple items MUST have plurality in the name. Ex: `truncate_all`, `keep_keys`, `replace_all_matches`.
- Functions that interact with a single item MUST NOT have plurality in the name. If a function would interact with multiple items due to a condition, like `where`, it is still considered singular. Ex: `set`, `delete`, `replace_match`.
- Functions that change a specific target MUST set the target as the first parameter.

## Adding New Editors/Converters

> Before raising a PR with a new function or converter, raise a PR to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.

Reviewer 3: ```suggestion
Before raising a PR with a new function or converter, raise an Issue to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.
```

- Functions that change a specific target MUST set the target as the first parameter.

## Adding New Editors/Converters

Before raising a PR with a new function or converter, raise an Issue to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.

Your proposal likely will be accepted if:
- The proposed functionality is missing,
- The proposed solution significantly improves user experience and readability for very common use cases,
- The proposed solution is more performant in cases where it is possible to achieve the same result with existing options.

It will be up for discussion if your proposal solves an issue that can be achieved in another way but does not improve user experience or performance.

Your proposal likely won't be accepted if:
- User experience is worse and assumes a highly technical user,
- The performance of your proposal very negatively affects the processing pipeline.

> As with code, `ottl` should aim for readability first. This means:

Reviewer: ```suggestion
As with code, OTTL aims for readability first. This means:
```
